Add GLPNImageProcessorFast

Aravind-11 opened the pull request:
# What does this PR do?

This PR adds a **fast image processor for the GLPN model**, implemented as `GLPNImageProcessorFast`.  

Fixes # (issue)

## Before submitting
- Implements `GLPNImageProcessorFast` using `BaseImageProcessorFast`.
- Adds tests and documentation updates.

### 🧪 Testing
- All tests pass except for the (`test_slow_fast_equivalence_batched`). I would like some help here. 

### 📄 Files updated
- `src/transformers/models/glpn/image_processing_glpn_fast.py`
- `src/transformers/models/glpn/__init__.py`
- `src/transformers/models/auto/image_processing_auto.py`
- `tests/models/glpn/test_image_processing_glpn.py`
- `docs/source/en/model_doc/glpn.md`

## Before submitting
- [x] Read the [contributor guidelines](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request).
- [x] Updated documentation and tests.
- [x] Verified style and quality with `make style` and `make quality`.

## Who can review?

@yonigozlan @molbap

Aravind-11 committed:
Add GLPNImageProcessorFast for torch backend

molbap commented on src/transformers/models/glpn/image_processing_glpn_fast.py:
```diff
@@ -0,0 +1,230 @@
+    # Back-compat alias (NOT persisted):
+    ensure_multiple_of: int
+    # Allow overriding resample (persisted like slow):
+    resample: PILImageResampling
```
Why isn't this imported?

molbap reviewed (commented):
Hey, thanks for starting this! Left some initial comments :)

Aravind-11 committed:
Address review feedback

- Simplified to_dict() method
- Keep tensors as torch instead of converting to numpy for heterogeneous shapes
- Removed unnecessary shape guards in post_process_depth_estimation
- Improved variable names (tgt -> target_size, d -> resized)
- Removed unnecessary GLPNImageProcessorKwargs class

Aravind-11 committed:
Address review feedback

- Simplified to_dict() method
- Keep tensors as torch instead of converting to numpy for heterogeneous shapes
- Removed unnecessary shape guards in post_process_depth_estimation
- Improved variable names (tgt -> target_size, d -> resized)
- Removed unnecessary GLPNImageProcessorKwargs class

Aravind-11 commented on tests/models/glpn/test_image_processing_glpn.py:
```diff
@@ -109,55 +125,70 @@ def test_image_processor_properties(self):
     def test_call_numpy_4_channels(self):
-        # Initialize image_processing
         image_processing = self.image_processing_class(**self.image_processor_dict)
-        # create random numpy tensors
```
Added them back.

Aravind-11 commented:
> Hey, thanks for starting this! Left some initial comments :)

Thanks a lot for reviewing Pablo! I've made the changes.

molbap commented on src/transformers/models/glpn/image_processing_glpn_fast.py:
```diff
@@ -0,0 +1,199 @@
+    # Don't persist an explicit `size` for GLPN (slow doesn't)
+    image_mean = IMAGENET_STANDARD_MEAN
+    image_std = IMAGENET_STANDARD_STD
+    size = {"height": 480, "width": 640}  # only for validation; we still crop, not resize
```
ah but size is actually defined here - no need to re-define it after!

molbap reviewed (commented):
Thanks for iterating! Did a second review 🤗

Aravind-11 commented on src/transformers/models/glpn/image_processing_glpn_fast.py:
```diff
@@ -0,0 +1,199 @@
+            if key not in keys_to_keep:
+                d[key] = None
+
+        return d
```
Ahh my bad! Sorry.

Aravind-11 commented on src/transformers/models/glpn/image_processing_glpn_fast.py:
```diff
@@ -0,0 +1,199 @@
+    image_std = IMAGENET_STANDARD_STD
+    size = {"height": 480, "width": 640}  # only for validation; we still crop, not resize
+    interpolation = F.InterpolationMode.BILINEAR
+    # valid_kwargs = GLPNImageProcessorKwargs
```
I defined the kwargs in the slow processor and imported it.

Aravind-11 committed:
commits after 2nd review

Aravind-11 committed:
Address all review feedback and add explicit batched test

- Simplified to_dict() with descriptive variable names (d->output_dict)
- Fixed resize operation: changed from crop to proper resize with interpolation
- Added padding for heterogeneous batch shapes in both slow and fast processors
- Fused rescale and normalize operations for efficiency
- Improved all variable names (tgt->target_size, d->depth_4d->resized)
- Added GLPNImageProcessorKwargs class in slow processor and imported in fast
- Renamed test_equivalence_slow_fast to test_slow_fast_equivalence
- Added explicit test_slow_fast_equivalence_batched test
- All 20 tests passing

molbap reviewed (commented):
I left additional comments because I'm not 100% convinced by the padding logic, let's make sure it's needed, and if it is let's use existing methods!

Aravind-11 committed:
using padding from utils

Aravind-11 committed:
Merge branch 'main' into add-glpn-fast-processor

yonigozlan committed:
simplify glpn image processor fast

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into add-glpn-fast-processor

yonigozlan reviewed (approved):
Hey @Aravind-11, thanks a lot for working on this! I made some final changes to get this merged. Mostly removed the padding logic so as not to break BC as it wasn't in the original image processor.
I'll merge when the CI passes!

yonigozlan committed:
fix docstring

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_41725). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

yonigozlan committed:
Merge branch 'main' into add-glpn-fast-processor

Aravind-11 committed:
Merge branch 'main' into add-glpn-fast-processor

yonigozlan committed:
Merge branch 'main' into add-glpn-fast-processor
